contracts/RelayHub.sol

```diff
@@ -135,19 +142,24 @@ contract RelayHub is IRelayHub {
             IPaymaster(relayRequest.relayData.paymaster).getGasAndDataLimits{gas:50000}();
         require(msg.data.length <= gasAndDataLimits.calldataSizeLimit, "msg.data exceeded limit" );
         uint256 dataGasCost = calldataGasCost(msg.data.length);
+        uint256 txDataCost = externalGasLimit - initialGasLeft - relayCallDataOverhead;
+        uint256 txDataCostPerByte = txDataCost/msg.data.length;
+        require(txDataCostPerByte >= G_ZERO && txDataCostPerByte <= G_NONZERO, "invalid externalGasLimit");
+
         require(maxRelayExposure >= gasAndDataLimits.acceptanceBudget.add(dataGasCost), "pm budget + dataGasCost too high");
 
         maxPossibleGas =
             gasOverhead.add(
             gasAndDataLimits.preRelayedCallGasLimit).add(
             gasAndDataLimits.postRelayedCallGasLimit).add(
             relayRequest.request.gas).add(
-            dataGasCost);
+            dataGasCost).add(
+            txDataCost);
 
         // This transaction must have enough gas to forward the call to the recipient with the requested amount, and not
         // run out of gas later in this function.
         require(
-            initialGas >= maxPossibleGas,
+            externalGasLimit >= maxPossibleGas,
```

Comment:
Maybe if we are touching the line, this is a good place to add another `gasLimit * (63/64)^3` check? Just asking. It is on relay's account if the gas is insufficient in most cases

Reply:
its a separate issue, and requires its own test (the change is simple, but the test isn't..)

Reply:
(also, didn't really touch it: I renamed the "initialGas" parameter to its proper name, "externalGasLimit", since there was no point giving it a new name..